### PR TITLE
8265393: VM crashes if both -XX:+RecordDynamicDumpInfo and -XX:SharedArchiveFile options are specified

### PR DESCRIPTION
**Pull request by yminqi**
Hi, Please review
  1) When the two flags -XX:+RecordDynamicDumpInfo and -XX:SharedArchiveFile used,  in
     Arguments::init_shared_archive_paths() :
          if (DynamicDumpSharedSpaces) {
            if (os::same_files(SharedArchiveFile, ArchiveClassesAtExit)) {
               ...
  ArchiveClassesAtExit is NULL.
  C++ does not have a defined behavior of strcmp with strings of NULL. Posix version of os::same_files(const char* file1, const char* file2) calls strcmp without checking input args. This should be same as it is treated on Windows.
  2) JCmdTest.java is fat and possible to cause test timeout. Split JCmdTest.java into two separate Tests: JCmdTestStaticDump.java and JCmdTestDynamicDump.java.
  3) Add a test to check run with -XX:+RecordDynamicDumpInfo and -XX:SharedArchiveFile in JCmdTestDynamicDump.java
       The test does not use utility to create jvm process  since it will take testing envs into process, but the way in the test using LingeredApp does not take envs.

  Tests: tier1,tier2,tier3,tier4

Thanks
Yumin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265393](https://bugs.openjdk.java.net/browse/JDK-8265393): VM crashes if both -XX:+RecordDynamicDumpInfo and -XX:SharedArchiveFile options are specified

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3599/head:pull/3599` \
`$ git checkout pull/3599`

Update a local copy of the PR: \
`$ git checkout pull/3599` \
`$ git pull https://git.openjdk.java.net/jdk pull/3599/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3599`

View PR using the GUI difftool: \
`$ git pr show -t 3599`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3599.diff">https://git.openjdk.java.net/jdk/pull/3599.diff</a>

</details>

### CODE REVIEW

**Review by iklam — CHANGES_REQUESTED** [review]
There's duplicated code between JCmdTestDynamicDump.java and JCmdTestStaticDump.java. I think these should be moved to a common base class.

**Review by calvinccheung — CHANGES_REQUESTED** [review]
Looks good. I spotted some minor issues in the test code.

Thanks,
Calvin

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.cds.CDSTestUtils;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java:31]: Extra import

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import jdk.test.lib.dcmd.PidJcmdExecutor;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.JDKToolFinder;
+import jtreg.SkippedException;
+import sun.hotspot.WhiteBox;
+
+import java.io.InputStreamReader;
+import java.io.BufferedReader;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java:40]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    public static void runTest(JCmdTest t) throws Exception {
+        checkCDSEnabled();
+        t.run();
+    }
+    private static final String TEST_CLASSES[] =
+                             {"JCmdTestLingeredApp",
+                              "jdk/test/lib/apps/LingeredApp",
+                              "jdk/test/lib/apps/LingeredApp$1"};
+    private static final String BOOT_CLASSES[]    = {"Hello"};
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java:71]: Too many blank spaces before the '='.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    protected static String testJar = null;
+    protected static String bootJar = null;
+    protected static String allJars = null;
+
+    protected static void buildJars() throws Exception {
+        testJar = JarBuilder.build("test", TEST_CLASSES);
+        bootJar = JarBuilder.build("boot", BOOT_CLASSES);
+        System.out.println("Jar file created: " + testJar);
+        System.out.println("Jar file created: " + bootJar);
+        allJars = testJar+ File.pathSeparator + bootJar;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDumpBase.java:82]: Need a blank space after `testJar`.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ * @build sun.hotspot.WhiteBox
+ * @build JCmdTestLingeredApp JCmdTestDynamicDump
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm/timeout=480 -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI JCmdTestDynamicDump
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:42]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.cds.CDSTestUtils;
+import jdk.test.lib.dcmd.PidJcmdExecutor;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:46]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.JDKToolFinder;
+
+import java.io.InputStreamReader;
+import java.io.BufferedReader;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestDynamicDump.java:52]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+ * @build JCmdTestLingeredApp JCmdTestStaticDump
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm/timeout=480 -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI JCmdTestStaticDump
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java:43]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.dcmd.PidJcmdExecutor;
+import jdk.test.lib.process.OutputAnalyzer;
+
+import java.io.InputStreamReader;
+import java.io.BufferedReader;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java:50]: Extra import.

File: test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    private static final int ITERATION_TIMES = 2;
+
+    static void test() throws Exception {
+        setIsStatic(true);
+        buildJars();
+        int  test_count = 1;
+        final boolean useBoot = true;
+        final boolean noBoot = !useBoot;
+        final boolean EXPECT_PASS = true;
+        final boolean EXPECT_FAIL = !EXPECT_PASS;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTestStaticDump.java:89]: These are also used in JCmdTestDynamicDump. Maybe move them to JCmdTestDumpBase?

**Review by calvinccheung — APPROVED** [review]
LGTM.

**Comment by yminqi** [comment]
> LGTM.

Thanks for review!

**Review by iklam — APPROVED** [review]
The latest code looks good to me.

**Comment by yminqi** [comment]
@iklam Thanks for review.

**Pull request closed without merging**